File: samplecode/node/naive.js
#!/usr/bin/env nodejs

//	  Copyright 2014 athenahealth, Inc.
//
//	 Licensed under the Apache License, Version 2.0 (the "License"); you
//	 may not use this file except in compliance with the License.  You
//	 may obtain a copy of the License at
//
//		 http://www.apache.org/licenses/LICENSE-2.0
//
//	 Unless required by applicable law or agreed to in writing, software
//	 distributed under the License is distributed on an "AS IS" BASIS,
//	 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
//	 implied.  See the License for the specific language governing
//	 permissions and limitations under the License.


var events = require('events')
var https = require('https')
var querystring = require('querystring')

var key = 'CHANGEME: YOUR_API_KEY'
var secret = 'CHANGEME: YOUR_API_SECRET'
var version = 'preview1'
var practiceid = 000000

var auth_prefixes = {
	v1: '/oauth',
	preview1: '/oauthpreview',
	openpreview1: '/oauthopenpreview',
}

// This is a useful function to have
function path_join() {
	// trim slashes from arguments, prefix a slash to the beginning of each, re-join (ignores empty parameters)
	var args = Array.prototype.slice.call(arguments, 0)
	var nonempty = args.filter(function(arg, idx, arr) {
		return typeof(arg) != 'undefined'
	})
	var trimmed = nonempty.map(function(arg, idx, arr) {
		return '/' + String(arg).replace(new RegExp('^/+|/+$'), '')
	})
	return trimmed.join('')
}

// Since we want these functions to run in a set order, we need a way to signal for the next one.
var signal = new events.EventEmitter

// We need to save the token in an outer scope, because of callbacks.
var token

function authentication() {
	var req = https.request({
		// Set up the request, making sure the content-type header is set. Let the https library do
		// the auth header (including base64 encoding) for us.
		hostname: 'api.athenahealth.com',
		method: 'POST',
		path: path_join(auth_prefixes[version], '/token'),
		auth: key + ':' + secret,
		headers: {'content-type': 'application/x-www-form-urlencoded'},
	}, function(response) {
		response.setEncoding('utf8')
		var content = ''
		response.on('data', function(chunk) {
			content += chunk
		})
		response.on('end', function() {
			var authorization = JSON.parse(content)
			// Save the token!
			token = authorization.access_token
			console.log(token)
			signal.emit('next')
		})
	})

	req.on('error', function(e) {
		console.log(e.message)
	})

	// The one parameter required for OAuth
	req.write(querystring.stringify({grant_type: 'client_credentials'}))
	req.end()
}

function departments() {
	// Create and encode parameters
	var parameters = {
		limit: 1,
	}
	var query = '?' + querystring.stringify(parameters)

	var req = https.request({
		hostname: 'api.athenahealth.com',
		method: 'GET',
		path: path_join(version, practiceid, '/departments', query),
		// We set the auth header ourselves this time, because we have a token now.
		headers: {'authorization': 'Bearer ' + token},
	}, function(response) {
		response.setEncoding('utf8')
		var content = ''
		response.on('data', function(chunk) {
			content += chunk
		})
		response.on('end', function() {
			console.log('Department:')
			console.log(JSON.parse(content))
			signal.emit('next')
		})
	})
	req.on('error', function(e) {
		console.log(e.message)
	})

	req.end()
}

function notes() {
	var parameters = {
		notetext: 'Javascript says hi!',
	}
	var content = querystring.stringify(parameters)
	var req = https.request({
		hostname: 'api.athenahealth.com',
		method: 'POST',
		path: path_join(version, practiceid, '/appointments/1/notes'),
		headers: {
			'authorization': 'Bearer ' + token,
			'content-type': 'application/x-www-form-urlencoded',
			'content-length': content.length, // apparently we have to set this ourselves when using
											  // application/x-www-form-urlencoded
		},
	}, function(response) {
		response.setEncoding('utf8')
		var content = ''
		response.on('data', function(chunk) {
			content += chunk
		})
		response.on('end', function() {
			console.log('Note posted:')
			console.log(JSON.parse(content))
			signal.emit('next')
		})
	})
	req.on('error', function(e) {
		console.log(e.message)
	})

	req.write(content)
	req.end()
}

// This is one way of forcing the call order
function main() {
	var calls = [authentication, departments, notes]
	signal.on('next', function() {
		var nextCall = calls.shift()
		if (nextCall) {
			nextCall()
		}
	})
	signal.emit('next')
}

main()

File: samplecode/python/naive.py
#!/usr/bin/env python2

#    Copyright 2014 athenahealth, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License"); you
#   may not use this file except in compliance with the License.  You
#   may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#   implied.  See the License for the specific language governing
#   permissions and limitations under the License.

import httplib
import base64
import json
import urllib

key = 'CHANGEME: YOUR_API_KEY'
secret = 'CHANGEME: YOUR_API_SECRET'
practiceid = 000000
version = 'preview1'

# Start a connection
connection = httplib.HTTPSConnection('api.athenahealth.com')


# Authenticate (basic access authentication)
auth_prefixes = {
	'v1': '/oauth',
	'preview1': '/oauthpreview',
	'openpreview1': '/oauthopenpreview',
}
auth_path = auth_prefixes[version] + '/token'

keypair = base64.b64encode('{0}:{1}'.format(key, secret))
auth_parameters = urllib.urlencode({'grant_type': 'client_credentials'})
auth_headers = {
	'Content-type': 'application/x-www-form-urlencoded',
	'Authorization': 'Basic {0}'.format(keypair),
}
auth_path = auth_path

connection.request('POST', auth_path, auth_parameters, auth_headers)
response = connection.getresponse()
authentication = json.loads(response.read())
token = authentication['access_token']
print token

# This token will be used in the header of every other request, so make it in advance.
token_header = {'Authorization': 'Bearer {0}'.format(token)}

# The path for every request has a prefix based on API version and practice ID, so set that now.
base_path = '/{0}/{1}'.format(version, practiceid)


# GET /departments
path = base_path + '/departments'

# Make sure to urlencode the parameters so they're escaped properly
parameters = urllib.urlencode({
	'limit': 1,
})
headers = {}
headers.update(token_header)	# Make sure to include the token!

# For GET requests, the parameters go in the URL, not the body, so add them to the URL manually.
# Pass in an empty string for the body (third argument), since GETs don't use it anyway.
connection.request('GET', path + '?' + parameters, '', headers)
response = connection.getresponse()
departments = json.loads(response.read())
print departments


# POST /appointments/{appointmentid}/notes
appointmentid = 1
path = base_path + '/appointments/{0}/notes'.format(appointmentid)
parameters = urllib.urlencode({
	'notetext': 'Hello from Python',
})
headers = {
	'Content-Type': 'application/x-www-form-urlencoded',
}
headers.update(token_header)

# POST parameters go in the body
connection.request('POST', path, parameters, headers)
response = connection.getresponse()
post_status = json.loads(response.read())
note_successful = post_status['success']
print note_successful
